### ISSUE TITLE
Smart Card Redirect

### ISSUE THREAD

**Issue by Jack-Vaughn** [FreeRDP/FreeRDP#6705]
**Describe the Bug**
Smart card is not successfully redirected to Windows. Smart card works locally and shows correctly under ```pcsc_scan```

**To Reproduce**
1. Boot a freshly installed Windows 10 machine
2. Run ```xfreerdp /u:<user> /v:<ip> /smartcard```
3. Open 'Device Manager' on Windows
4. See there is no 'Card Readers' listed

**Expected Behavior**
Emulated card reader is shown under Windows Device Manager and smart card successfully redirects

**Screenshots**
N/A

**Application Details**
-FreeRDP Version - 2.2.0
-Command line used - ```xfreerdp /u:<user> /v:<ip> /smartcard```
-Buildconfig - ```This is FreeRDP version 2.2.0 (n/a)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=ON WITH_GSSAPI=OFF WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_ICU=OFF WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_MODULES=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VAAPI=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          RELWITHDEBINFO
CFLAGS:              -g -O2 -fdebug-prefix-map=/build/freerdp2-u6XKVf/freerdp2-2.2.0+git202007201338=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 9.3.0
Target architecture: x64```

-Server Side OS Version - Windows 10.0.19042 Build 19042 (have tried multiple versions and builds of Windows 10)

-Log output - Attached [log.txt](https://github.com/FreeRDP/FreeRDP/files/5766474/log.txt)

**Environment**
-OS - Ubuntu
-Version - 20.04.1
-Architecture - amd64

**Additional Context**

- stdout
```
[17:35:08:824] [52329:52330] [INFO][com.freerdp.core] - freerdp_connect:freerdp_set_last_error_ex resetting error state

[17:35:08:824] [52329:52330] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr

[17:35:08:824] [52329:52330] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd

[17:35:08:824] [52329:52330] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr

[17:35:09:138] [52329:52330] [INFO][com.freerdp.primitives] - primitives autodetect, using optimized

[17:35:09:139] [52329:52330] [INFO][com.freerdp.core] - freerdp_tcp_is_hostname_resolvable:freerdp_set_last_error_ex resetting error state

[17:35:09:139] [52329:52330] [INFO][com.freerdp.core] - freerdp_tcp_connect:freerdp_set_last_error_ex resetting error state

[17:35:09:148] [52329:52330] [WARN][com.freerdp.crypto] - Certificate verification failure 'unable to get local issuer certificate (20)' at stack position 0

[17:35:09:148] [52329:52330] [WARN][com.freerdp.crypto] - CN = <computer-name>

[17:35:10:349] [52329:52330] [INFO][com.freerdp.gdi] - Local framebuffer format PIXEL_FORMAT_BGRX32

[17:35:10:349] [52329:52330] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16

[17:35:10:366] [52329:52330] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem

[17:35:10:366] [52329:52330] [INFO][com.freerdp.channels.rdpsnd.client] - [static] Loaded fake backend for rdpsnd

[17:35:10:366] [52329:52344] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)

[17:35:10:370] [52329:52344] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)

[17:35:10:709] [52329:52330] [INFO][com.freerdp.core] - rdp_set_error_info:freerdp_set_last_error_ex resetting error state

[17:35:12:219] [52329:52344] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)

[17:35:12:227] [52329:52344] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)

[17:35:12:463] [52329:52381] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetDeviceTypeIdW failed with error SCARD_E_UNSUPPORTED_FEATURE [-2146435038]

[17:35:12:463] [52329:52381] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetDeviceTypeId (0x00090108), status: SCARD_E_UNSUPPORTED_FEATURE (0x80100022)

[17:35:12:466] [52329:52383] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetDeviceTypeIdW failed with error SCARD_E_UNSUPPORTED_FEATURE [-2146435038]

[17:35:12:466] [52329:52383] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetDeviceTypeId (0x00090108), status: SCARD_E_UNSUPPORTED_FEATURE (0x80100022)

[17:35:12:470] [52329:52385] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetDeviceTypeIdW failed with error SCARD_E_UNSUPPORTED_FEATURE [-2146435038]

[17:35:12:470] [52329:52385] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetDeviceTypeId (0x00090108), status: SCARD_E_UNSUPPORTED_FEATURE (0x80100022)
```

-```pcsc_scan``` output
```
Using reader plug'n play mechanism

Scanning present readers...

0: SCM Microsystems Inc. SCR 3310 [CCID Interface] (xxxx) 00 00

Wed Dec 30 17:36:18 2020

Reader 0: SCM Microsystems Inc. SCR 3310 [CCID Interface] (xxxx) 00 00

Event number: 0

Card state: Card inserted,

ATR: xxxx



ATR: xxxx

+ TS = 3B --> Direct Convention

+ T0 = F9, Y(1): 1111, K: 9 (historical bytes)

TA(1) = 18 --> Fi=372, Di=12, 31 cycles/ETU

129032 bits/s at 4 MHz, fMax for Fi = 5 MHz => 161290 bits/s

TB(1) = 00 --> VPP is not electrically connected

TC(1) = 00 --> Extra guard time: 0

TD(1) = 00 --> Y(i+1) = 0000, Protocol T = 0

-----

+ Historical bytes: xxxx

Category indicator byte: 53 (proprietary format)



Possibly identified card (using /home/<user>/.cache/smartcard_list.txt):

xxxx

G+D FIPS 201 SCE 7.0 (PKI)
```

**Comment by auron-muc** [comment]
Similar problem here, slightly different setup.

Using freerdp-nightly-3.0.0~20210103... from pub.freerdp.com repo on Ubuntu 20.04 to connect to a windows 10.2004 machine with all patches. 

Using opensc 0.21 (both x64 and 32 installed) on windows. 

Everything is fine, except the smartcard redirect. 

Output from opensc-tool on linux:

`#user:~$ opensc-tool -l -v -v`
`# Detected readers (pcsc)`
`Nr.  Card  Features  Name`
`0    Yes             Nitrokey Nitrokey HSM (<SERIAL_REDACTED>         ) 00 00`
`     3b:de:18:ff:81:91:fe:1f:c3:80:31:81:54:48:53:4d:31:73:80:21:40:81:07:1c SmartCard-HSM version 3.3 `
`#user:~$ opensc-tool -n -v -v`
`Using reader with a card: Nitrokey Nitrokey HSM (<SERIAL_REDACTED>         ) 00 00`
`Connecting to card in reader Nitrokey Nitrokey HSM (<SERIAL_REDACTED>         ) 00 00...`
`P:19762; T:0x140322067396672 16:02:02.432 [opensc-tool] card.c:254:sc_connect_card: called`
`P:19762; T:0x140322067396672 16:02:02.433 [opensc-tool] card-entersafe.c:138:entersafe_match_card: called`
`P:19762; T:0x140322067396672 16:02:02.433 [opensc-tool] card-rutoken.c:103:rutoken_match_card: called`
`P:19762; T:0x140322067396672 16:02:02.433 [opensc-tool] card-coolkey.c:2240:coolkey_match_card: called`
`P:19762; T:0x140322067396672 16:02:02.433 [opensc-tool] card-coolkey.c:919:coolkey_apdu_io: called`
`P:19762; T:0x140322067396672 16:02:02.433 [opensc-tool] card-coolkey.c:2385:coolkey_card_reader_lock_obtained: called`
`P:19762; T:0x140322067396672 16:02:02.486 [opensc-tool] card-muscle.c:832:muscle_card_reader_lock_obtained: called`
`P:19762; T:0x140322067396672 16:02:02.522 [opensc-tool] muscle.c:276:msc_select_applet: returning with: -1200 (Card command failed)`
`Using card driver SmartCard-HSM.`
`Card name: SmartCard-HSM version 3.3`
`P:19762; T:0x140322067396672 16:02:02.572 [opensc-tool] ctx.c:927:sc_release_context: called`

Output from Opensc-tool on Windows:
`PS C:\Program Files (x86)\OpenSC Project\OpenSC\tools> .\opensc-tool -l -v -v`
`# Detected readers (pcsc)`
`Nr.  Card  Features  Name`
`0    No              Nitrokey Nitrokey HSM (<SERIAL_REDACTED>         ) 00 00`
`PS C:\Program Files (x86)\OpenSC Project\OpenSC\tools> .\opensc-tool -n -v -v`
`Card not present.`
`Failed to connect to reader: Card not present`
`P:12744; T:7828 2021-01-05 17:16:03.034 [opensc-tool] ctx.c:931:sc_release_context: called`

So, the card reader is seen on windows, but it cannot detect the card. (Card = "No" on Windows vs. Card = "Yes" on Linux)
I can't see the reader in device manager either, same as OP.

freerdp segfaults pretty immediately after connection when I try to run with /log-level:TRACE

Should I open a separate Issue, or are these related?

**Comment by akallabeth** [comment]
@auron-muc please create a new issue if you have any segfault (pleas attach as much information as possible)
As for the smartcard part, you can try with `/client-build-number:2600` and hope it works. The issue you describe is that the RDP smartcard redirection requests a feature that can not be mapped to `pcscd`.
That would require some patches there first before we can support that.

**Comment by Jack-Vaughn** [comment]
I tried upgrading pcscd to 1.9.0 with no luck. Using option ```/client-build-number:2600``` did not work either.

**Comment by akallabeth** [comment]
@Jack-Vaughn then you´re currently out of luck.
The issue is not really a bug in `pcscd` or `freerdp` bug a mismatch in APIs for smartcards between windows and others.
RDP (basically) only redirects the windows syscalls and we can not implement some of them as `pcscd` does not provide the required information.
The `/client-build-number:2600` forces older smart card API use (don´t ask why this flag, this hack is part of the core protocol...)

**Comment by akallabeth** [comment]
Could you give it another try? #6792 fixes lots of smartcard related stuff

**Comment by Jack-Vaughn** [comment]
This resolved the problem!
